Implement master controller

staebler opened the pull request:
**Changes**
* Implement the controller that is responsible for installing openshift on the master nodes.
* Create a JobSync type to pull together more of the duplicated code around syncing for controllers that rely on jobs.
* Increase the duration that ansible jobs can run before failing. This is needed now because the machines created in AWS are not available for a while after the scale groups are created. The machineset job that creates the scale groups completes successfully prior to the machines in the scale groups being accessible. The pods of the master job will fail until the master machines are accessible. We wanted to increase this duration anyway, so now is as good a time as any.

**Notes**
* The latest docker.io/openshift/origin-ansible does not work with cluster-operator. @abutcher has a PR up to fix at least one of the problem. In the meantime, you can build your own openshift-ansible image to use. I tested with commit 3346448363ead62312cb29596dcf3df5ce753ef7.
* The master job is not passing for me. It is failing trying to ssh into the master machine. The logic for controlling the master is working. Given the size of this PR, I wanted to get it in even though the work to install on the master machines is not complete.

csrwng commented on pkg/controller/master/master_controller.go:
```diff
@@ -160,31 +221,31 @@ func (c *MasterController) Run(workers int, stopCh <-chan struct{}) {
-func (c *MasterController) enqueueAfter(machine *clusteroperator.Machine, after time.Duration) {
-	key, err := controller.KeyFunc(machine)
+// enqueueAfter will enqueue a machine set after the provided amount of time.
+func (c *MasterController) enqueueAfter(machineSet *clusteroperator.MachineSet, after time.Duration) {
```
Is there a case where we foresee using this?

staebler replied:
Yeah, I think this is what we would use when we run the period reconciliation that re-asserts that everything is set up in AWS.

csrwng replied:
sgtm

csrwng commented on pkg/controller/master/master_controller.go:
```diff
@@ -217,56 +278,215 @@ func (c *MasterController) handleErr(err error, key interface{}) {
+	}
+}
+
+func (c *MasterController) updateMachineSetStatus(original, machineSet *clusteroperator.MachineSet) error {
```
Should this be moved to controller_utils?

staebler replied:
Sure.

staebler commented on pkg/controller/controller_utils.go:
```diff
@@ -66,20 +52,34 @@ func WaitForCacheSync(controllerName string, stopCh <-chan struct{}, cacheSyncs
 	return true
 }
 
+type UpdateConditionCheck func(oldReason, oldMessage, newReason, newMessage string) bool
```
Ah, that's a deficiency. In that case, the condition should only be updated when the reason/message has changed. I need to fix this.

staebler commented on pkg/controller/jobsync.go:
```diff
@@ -0,0 +1,274 @@
+		return nil
+	// New job has not been created, so an old job must exist. Set the owner
+	// to not processing as the old job is deleted.
+	case job == nil:
```
Agreed about returning a more meaningful result. I briefly considered doing that for this PR but held off on it. I'll create an issue to track this to ensure it stays in my mind.

> Would we get to this case if expectations are not met yet? If so, are we doing the right thing by setting the owner to not processing?

This case happens when (1) there were no pending expectations and (2) there were existing jobs to delete. A new job was not created because we are waiting for the old jobs to be deleted first. The owner is set to not processing because a process job (that may have been active) was deleted.

csrwng replied:
> This case happens when (1) there were no pending expectations and (2) there were existing jobs to delete. A new job was not created because we are waiting for the old jobs to be deleted first. The owner is set to not processing because a process job (that may have been active) was deleted.

If we have not met expectations, s.jobControl.ControlJobs returns nil, false, nil ... when do we catch that?

staebler replied:
That would be caught by the `job == nil` case, setting the owner to no provisioning, since there is not an active job.

csrwng replied:
hmm, doesn't seem right. In the cluster controller we simply return when expectations have not been met yet

staebler replied:
Make sense. I'll re-work this to (1) make it more clear and (2) make it more correct.

staebler commented on pkg/controller/jobsync.go:
```diff
@@ -0,0 +1,274 @@
+		logger.Debugln("job exists, will sync with job")
+		return s.syncOwnerStatusWithJob(owner, job)
+	// Owner should have a job but it was not found.
+	case s.strategy.GetOwnerJobSyncConditionStatus(owner, JobSyncProcessing):
```
You're right. We don't seem to have a way to notice when an active job can no longer be found. If an active job is deleted outside of clust-op, the clust-op will just recreate the job. If I recall correctly, that fits the behavior that we agreed to the other week. Is that your recollection? If so, I'll delete this case and call it good.

csrwng replied:
Actually I think we decided the opposite :)
If I recall correctly, failing the provisioning/install when a job has been deleted is the only way that you can abort without deleting the owning resource. But yes, not sure how we'd detect that. For now, I'm ok with deleting the case and creating an issue to resolve it in the future.

staebler replied:
Yeah, that sounds right.

staebler commented:
@csrwng I've addressed all of your comments. There are significant changes to jobControl and jobSync. But I think it is much easier to follow now. I ended up adding back a `LocalObjectReference` to the current job so that we can track when the job may have been deleted out-of-band.

csrwng commented:
/lgtm
